sedaro/src/sedaro/branches/scenario_branch/sim_client.py

```diff
@@ -767,3 +788,58 @@ def produce(self, agent_id: str, external_state_id: str, values: tuple, timestam
                 **COMMON_API_KWARGS,
             )
         return tuple(serdes(v) for v in body_from_res(response))
+
+    def open_cosim(self):
+        api_key = self.__sim_client.get_sedaro()._api_key
+      
+        host = self.__sim_client.get_sedaro()._api_host
+
+        try:
+            address = self.__sim_client.status().get(
+                "clusterAddr")
+            job_id = self.__sim_client.status().get("id")
+            runner = grpc_client.CosimRunner()
+            runner.open(api_key, address, job_id, host, "localhost:50049")
+            self.__message_generator = runner.message_generator
+        except Exception as e:
+            raise e
+        
+    
+    def close_cosim(self):
+        if self.__message_generator:
+            self.__message_generator.terminate()
+            self.__message_generator = None
+        else:
+            raise Exception('No message generator has been set.   Open a cosimulation session first by calling `open_cosim`.')
+
+    def _consume_cosim(self, agent_id: str, external_state_id: str, time: float = None):
+        if not self.__message_generator:
+            raise Exception('No message generator has been set.   Open a cosimulation session first by calling `open_cosim`.')
+        else:
+            res = self.__message_generator.consume(external_state_id, agent_id, time)
+        return tuple(res)
+    
+
+    def _produce_cosim(self, agent_id: str, external_state_id: str, values: tuple, timestamp: float = None):
+        if not self.__message_generator:
+            raise Exception('No message generator has been set.   Open a cosimulation session first by calling `open_cosim`.')
+        else:
+            res = self.__message_generator.produce(
+                external_state_id, agent_id, values, timestamp)
+        return tuple(res)
+    
+    def consume_cosim(self, agent_id: str, external_state_id: str, time: float = None):
```

Comment:
`consume_cosim` isn't a great name.  We have the traditional REST `consume` and `produce` APIs.  Maybe the new APIs can be differentiated as streaming or async or both.